src/bootstrap/src/core/config/config.rs

```diff
@@ -2421,13 +2422,16 @@ impl Config {
         let compiler = format!("{top_level}/compiler/");
         let library = format!("{top_level}/library/");
 
+        // If we are running in CI, the current commit will never have artifacts already built.
+        let tip_commit = if CiEnv::is_ci() { "HEAD^" } else { "HEAD" };
```

Comment:
This feels wrong to me. The current commit *doesn't* have artifacts built, but that should never be relevant: we shouldn't ever use this commit:

1. If the compiler sources are unchanged in tip, in which case the commit we are downloading from should not be the tip (since the previous commit - or some earlier commit - is just as good and is more likely to be in cache already locally).
2. If the compiler sources *are* changed in tip, then I'd expect that to also mean that we can't use download-rustc -- after all, we're specifying if-unchanged, right? We need to build the compiler in that case.

It's possible that this conditional is still right, but if so, the justification in the comment above it is wrong. The problem is not that we don't have artifacts, it's (2) -- that we cannot use artifacts for the tip commit. But I've not traced the logic through bootstrap to be confident either way, so please do confirm.

Reply:
Hmm. Yes, this is incorrect. I just realized there is an edge case. So, if I make changes on compiler and push, this line will use the previous commit which will lead to an incorrect assumption about whether the compiler has changed. I will fix this, thank you.